### ISSUE TITLE
onclause is not accepted as keyword argument to Query.join

### ISSUE THREAD

**Issue by dtantsur** [sqlalchemy/sqlalchemy#4427]
When I tried joining two tables with cross-referencing foreign keys I got
```
>>> session.query(B).join(A)
...
  File "/home/dtantsur/Projects/temp/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 2361, in _join_left_to_right
    self._join_to_left(l_info, left, right, onclause, outerjoin, full)
  File "/home/dtantsur/Projects/temp/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 2496, in _join_to_left
    "Tried joining to %s, but got: %s" % (right, ae))
sqlalchemy.exc.InvalidRequestError: Could not find a FROM clause to join from.  Tried joining to <class '__main__.A'>, but got: Can't determine join between 'table_b' and 'table_a'; tables have more than one foreign key constraint relationship between them. Please specify the 'onclause' of this join explicitly.
```
So I tried
```
>>> session.query(B).join(A, onclause=(A.id == B.a_id))
Traceback (most recent call last):
  File "demo.py", line 44, in <module>
    session.query(B).join(A, onclause=(A.id == B.a_id))
  File "/home/dtantsur/Projects/temp/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 2133, in join
    ', '.join(sorted(kwargs)))
TypeError: unknown arguments: onclause
```
It seems that the `onclause` argument can only be passed from the ORM to the Core level as a positional argument:
```
>>> session.query(B).join(A, A.id == B.a_id)
```
This is pretty confusing because the Core `join` does have the `onclause` argument indeed. I suggest either the keyword form gets accepted or the error message is changed to refer to "the 2nd argument" instead of "the 'onclause'".

**Comment by zzzeek** [comment]
Hi Dmitry!

Unfortunately this is the [documented](https://docs.sqlalchemy.org/en/latest/orm/query.html?highlight=query%20join#sqlalchemy.orm.query.Query.join) behavior of query.join(), which also can accommodate for a chain of ON clauses, hence the arguments are passed positionally, such as:

   query = query.join(A.b, B.c, C.d)

now, the above calling style is also not something I am thrilled about, however it has different semantics than a string of three separate query.join() calls in some cases, which are also cases I consider to be somewhat "legacy" so are all on the chopping block for SQLAlchemy 2 which is where I plan to unify the workings of "join()" in Core and ORM, which have a lot of other differences too that I'm surprised is not what you actually hit.

One such case where query.join(a, b, c) is different than query.join(a).join(b).join(c) is, suppose "B" has an attribute "items", and A has an attribute "bs".  You can say this:

   query(A).join("bs", "items")

but this would fail:

   query(A).join("bs").join("items")

the way you can make that work, without changing the argument style, would be:

   query(A).join("bs").join("items", from_joinpoint=True)

now, in modern SQLAlchemy, nobody ever does that, they'd just do this:

   query(A).join(A.bs).join(B.items)

so this is all legacy behaviors that require that the overly open-ended calling style of join() be scaled back which is a long term goal.

The problem I *thought* you had, is *way* more of a critical issue that confuses people, which is that select.join() does something completely different than query.join(); it turns the "selectable" into a subquery, rather than appending the criteria.   there's a great semantic reason for that but when Query came along in version 0.3, it created a large disparity between calling styles.     The ORM's API evolved quite slowly over many years which grows a lot of artifacts that continuously need to be trimmed back.

towards the goal of unifying select.join() and query.join(), there will be one join() method that only appends a single criteria to the selectable being built.

query.join() *does* correctly raise a TypeError here, so to the extent that the API is a little overburdened, it did the best it could.

will target this at the 2.0.

**Comment by zzzeek** [comment]
SQLAlchemy 1.4 / 2.0 establishes Query as a legacy object.   in the new model, select() is used, and this accepts onclause as a keyword argument:

```

    s = Session(e)

    s.execute(
            select(B).join(A, onclause=A.id == B.a_id)
   ).scalars().all()

```

so the problem is solved for 1.4 / 2.0 if the new API as part of #5159 is used.   that's all done.

As for the "chaining of joins" in Query itself,  the deprecation of which would encourage a programming pattern that provides a fixed number of arguments to join() and therefore "onclause" can be a keyword argument, at least in theory, in Query.join itself, deprecation of this pattern is set up in #4705.   so that issue tracks the other end of the most we could do for this issue.

2.0's plan is to keep Query around as a legacy construct, a lot of the calling styles will be removed (like using strings for relationship names in loader options), but it's not clear how much from Query will actually be fully removed, as opposed to emitting warnings, because the whole thing is there for legacy codebases only and if people are migrating their code they should use select().

so overall the issue here is covered by #5159 and #4705 and for users that migrate to the new APi the "onclause" argument may be named as a keyword.
